### ISSUE TITLE
[Bug]: multiple calendar entries for "Restmüll" in city of Karlsruhe

### ISSUE THREAD

**Issue by baschdl78** [mampfes/hacs_waste_collection_schedule#3303]
### I Have A Problem With:

A specific source

### What's Your Problem

The Home Assistant Calendar tab shows multiple (three) entries for "Restmüll" on every day it is scheduled. The other waste types are listed correct with only one calendar entry on the calendar Tab
![Screenshot_20241227_123759_Home Assistant](https://github.com/user-attachments/assets/5096546b-8102-4bef-9cc2-e05c2e9798af)

### Source (if relevant)

karlsruhe_de

### Relevant Configuration

```YAML
waste_collection_schedule:
    sources:
    - name: karlsruhe_de
      args:
        street: Kriegsstraße 
        hnr: 146
```

### Checklist Source Error

- [x] Use the example parameters for your source (often available in the documentation) (don't forget to restart Home Assistant after changing the configuration)
- [X] Checked that the website of your service provider is still working
- [X] Tested my attributes on the service provider website (if possible)
- [X] I have tested with the latest version of the integration (master) (for HACS in the 3 dot menu of the integration click on "Redownload" and choose master as version)

### Checklist Sensor Error

- [X] Checked in the Home Assistant Calendar tab if the event names match the types names (if types argument is used)

### Required

- [X] I have searched past (closed AND opened) issues to see if this bug has already been reported, and it hasn't been.
- [X] I understand that people give their precious time for free, and thus I've done my very best to make this problem as easy as possible to investigate.

**Comment by baschdl78** [comment]
The triple entry is also in "the normal" statd of entity as seen here[

![Image](https://github.com/user-attachments/assets/4488b78e-f3e0-4c35-b08a-0205ff498e0a)

](url)

**Comment by cunners** [comment]
It looks fine for me using the provided address. Is this still an issue?

**Comment by baschdl78** [comment]
> It looks fine for me using the provided address. Is this still an issue?

Yes it is. Still having three entries in calendar. Issue occurs only for Restmüll.

**Comment by cunners** [comment]
The calendars are not visible on your screenshot so I can't be sure but is there multiple entries for the wcs calendar? There should only be one and it covers all the collections.

**Comment by baschdl78** [comment]
> The calendars are not visible on your screenshot so I can't be sure but is there multiple entries for the wcs calendar? There should only be one and it covers all the collections.

Yes there is only one wcs calendar. It is the only calendar Marken as Active in the Screenshot. All other calendar are deactivated.

**Comment by baschdl78** [comment]
Hi,
I have deleted and reinstalled the city of Karlsruhe hub in homeassistant. There are still triple "Restmüll" entries. It's in the calendar as in the created sensor entity.
Can' i support with some search. Should I look in the source code if there's the triple as a typo or something else?
Or could there be something wrong in my local configuration file for the city of Karlsruhe? Where can I look for traces of the error?

**Comment by baschdl78** [comment]
**Issue:** The Karlsruhe (AWB) API returns separate entries for every physical container ID registered at an address. This leads to duplicate entries in both the calendar and sensors (e.g., "Restmüll" appearing three times on the same day) if multiple bins are present.

**Proposed Solution:** Implement a deduplication logic within the get_data method of karlsruhe_de.py. By tracking unique (date, waste_type) combinations using a set(), we can ensure that identical waste types on the same day are merged into a single collection event before being returned.

Code karlsruhe_de.py (tested locally):
```
from datetime import datetime

import requests
import urllib3
from waste_collection_schedule import Collection  # type: ignore[attr-defined]
from waste_collection_schedule.service.ICS import ICS

# With verify=True the POST fails due to a SSLCertVerificationError.
# Using verify=False works, but is not ideal.
urllib3.disable_warnings(urllib3.exceptions.InsecureRequestWarning)


TITLE = "City of Karlsruhe"
DESCRIPTION = "Source for City of Karlsruhe."
URL = "https://www.karlsruhe.de/"
TEST_CASES = {
    "Östliche Rheinbrückenstraße 1": {
        "street": "Östliche Rheinbrückenstraße",
        "hnr": 1,
    },
    "Habichtweg 4": {"street": "Habichtweg", "hnr": 4},
    "Machstraße 5": {"street": "Machstraße", "hnr": 5},
    "Bernsteinstraße 10 ladeort 1": {
        "street": "Bernsteinstraße",
        "hnr": 10,
        "ladeort": 1,
    },
    "Bernsteinstraße 10 ladeort 2": {
        "street": "Bernsteinstraße",
        "hnr": 10,
        "ladeort": 2,
    },
}


ICON_MAP = {
    "Restmüll": "mdi:trash-can",
    "Bioabfall": "mdi:leaf",
    "Papier": "mdi:package-variant",
    "Wertstoff": "mdi:recycle",
    "Sperrmüllabholung": "mdi:wardrobe",
}


API_URL = "https://web{i}.karlsruhe.de/service/abfall/akal/akal_{year}.php"


PARAM_TRANSLATIONS = {
    "de": {
        "street": "Straße",
        "hnr": "Hausnummer",
        "ladeort": "Ladeort",
    },
    "en": {
        "street": "Street",
        "hnr": "House number",
        "ladeort": "Loading point",
    },
}


class Source:
    def __init__(self, street: str, hnr: str | int, ladeort: int | None = None):
        self._street: str = street
        self._hnr: str | int = hnr
        self._ladeort: int | None = ladeort
        self.ics = ICS()

    def fetch(self):
        now = datetime.now()
        error = None
        for year in (now.year, now.year + 1, now.year - 1):
            for i in (4, 6):
                try:
                    return self.get_data(API_URL.format(year=year, i=i))
                except Exception as e:
                    error = e
        raise error

    def get_data(self, url):
        data = {
            "strasse_n": self._street,
            "hausnr": self._hnr,
            "ical": "+iCalendar",
            "ladeort": self._ladeort,
        }
        params = {"hausnr": self._hnr}

        r = requests.post(url, data=data, params=params, verify=False)
        dates = self.ics.convert(r.text)

        entries = []
        seen = set()  # Deduplication set
        
        for d in dates:
            date, waste_type = d
            # Karlsruhe adds details after a comma, we strip that for the UI
            waste_type = waste_type.split(",")[0].strip()
            
            # Create a unique identifier for this collection event
            identifier = (date, waste_type)
            
            if identifier not in seen:
                seen.add(identifier)
                icon = ICON_MAP.get(waste_type)
                entries.append(Collection(date=date, t=waste_type, icon=icon))

        return entries
```
**Testing Results:** I have applied this change to my local instance. It successfully reduced the triple "Restmüll" entries to a single, clean entry in the Home Assistant Calendar and sensors, without affecting other waste types or unique collection dates.

**Comment by baschdl78** [comment]
Update: The "error" really was on my end. An—obviously—incomplete uninstallation of a previous instance corrupted my configuration and generated the duplicate entries for the general waste. Uninstalling, COMPLETELY deleting the configuration directory, and then reinstalling did the trick.
